db/auth_service.go

```
@@ -0,0 +1,87 @@
+package db
+
+import (
+	"catalyst/internal"
+	"errors"
+
+	"gorm.io/gorm"
+)
+
+var (
+	DuplicateUsernameErr = errors.New("username is already registered")
+	DuplicateEmailErr    = errors.New("email is already registered")
+	BadRequestBodyErr    = errors.New("all fields are required")
+	InternalServerErr    = errors.New("internal server error")
+	IncorrectPasswordErr = errors.New("password is not correct")
+	IncorrectEmailErr    = errors.New("email is not correct")
+)
+
+// GetUserByEmail fetches a user from the database based on the provided email.
+func (d *DBClient) GetUserByEmail(email string) (User, error) {
+	var user User
+	result := d.DB.First(&user, "email = ?", email)
+	return user, result.Error
+}
+
+// GetUserByUsername fetches a user from the database based on the provided username.
+func (d *DBClient) GetUserByUsername(userName string) (User, error) {
+	var user User
+	result := d.DB.First(&user, "username = ?", userName)
+	return user, result.Error
+}
+
+// CreateUser adds a new user to the database.
+func (d *DBClient) CreateUser(user User) (User, error) {
+	result := d.DB.Create(&user)
+	return user, result.Error
+}
+
+// SignUp registers a new user if the email or username is not already in use.
+// It hashes the password and stores the user in the database.
+func (d *DBClient) SignUp(user User) (User, error) {
+	_, err := d.GetUserByEmail(user.Email)
+	if err == nil {
+		return User{}, DuplicateEmailErr
```

i think it's better to return the error directly and check the type of error in the endpoint handler.